src/NITTA/LuaFrontend.hs

```diff
@@ -1,474 +1,391 @@
-    put
-        alg
-            { algItems = Renamed{rFrom, rTo} : patch algItems
-            , algVars = rTo : algVars
-            }
+            return resultName
+addConstant _ = undefined
+
+addVariable :: (MonadState (AlgBuilder s t) m, IsString b) => [T.Text] -> [T.Text] -> [t] -> [Char] -> [Int] -> m b
+addVariable fIn fOut fValues fName fInt = do
+    AlgBuilder{algBuffer} <- get
+    let luaValueVersions = map (\x -> nameToLuaValueVersion algBuffer x) fOut
+    let func = Func{fIn, fValues, fName, fInt, fOut = luaValueVersions}
+    mapM_ (uncurry addItemToBuffer) $ zip fOut luaValueVersions
+    mapM_ addItemToVars luaValueVersions
+    algBuilder@AlgBuilder{algGraph} <- get
+    put algBuilder{algGraph = func : algGraph}
+    return $ getUniqueLuaVariableName (head luaValueVersions) 0
+    where
+        nameToLuaValueVersion algBuffer name =
+            case getLuaValueByName name algBuffer of
+                Just lvv@LuaValueVersion{luaValueVersionAssignCount} -> lvv{luaValueVersionAssignCount = luaValueVersionAssignCount + 1}
+                Nothing -> LuaValueVersion{luaValueVersionName = name, luaValueVersionAssignCount = 0, luaValueVersionIsConstant = False}
+        addItemToBuffer name lvv = do
+            algBuilder@AlgBuilder{algBuffer} <- get
+            put algBuilder{algBuffer = HashMap.insert name lvv algBuffer}
+        addItemToVars name = do
+            algBuilder@AlgBuilder{algVars} <- get
+            put algBuilder{algVars = HashMap.insert name [] algVars}
+
+addVariableAccess name = do
+    algBuilder@AlgBuilder{algVars} <- get
+    luaValueVersion <- getLatestLuaValueVersionByName name
+    case HashMap.lookup luaValueVersion algVars of
+        Just value -> do
+            let len = length value
+            let resultName = getUniqueLuaVariableName luaValueVersion len
+            put algBuilder{algVars = HashMap.insert luaValueVersion (resultName : value) algVars}
+            return resultName
+        Nothing -> error ("variable '" ++ show (luaValueVersionName luaValueVersion) ++ " not found. Constants list : " ++ show algVars)
+
+getLatestLuaValueVersionByName name = do
+    AlgBuilder{algBuffer} <- get
+    case HashMap.lookup name algBuffer of
+        Just value -> return value
+        Nothing -> error $ "variable not found : '" ++ show name ++ "'."
+
+addAlias from to = do
+    algBuilder@AlgBuilder{algBuffer} <- get
+    case getLuaValueByName to algBuffer of
+        Just value -> do
+            put algBuilder{algBuffer = HashMap.insert from value algBuffer}
+            return $ fromString ""
+        Nothing -> error ("variable '" ++ show to ++ " not found. Constants list : " ++ show algBuffer)
+
+getLuaValueByName name buffer = HashMap.lookup name buffer
+
+buildAlg syntaxTree =
+    flip execState st $ do
+        _ <- addStartupFuncArgs startupFunctionCall startupFunctionDef
+        mapM_ (processStatement startupFunctionName) $ funAssignStatements startupFunctionDef
```

Comment:
I think if you inline `funAssignStatements` to `let` or `do` statement, it will be more readable. Current naming doesn't allow us to understand that it is a just a getter.